### ISSUE TITLE
Generate wasm dynamically?

### ISSUE THREAD

**Issue by justinmchase** [pepyakin/wabt-rs#49]
Can you recommend a good way to generate wasm dynamically from an ast?

I want to make a language that produces an ast which I then want to compile into wasm. I don't want to use an existing language like C or Rust to statically compile it but I want to do it dynamically and as cleanly and directly as wat.

I could theoretically use my ast to generate wat and then run it through this api but I was hoping to skip the wat generation step. I was imagining that this library had a `ast2wasm` step but I don't think it does.

Do you have any recommendations on this approach before I go any further?

**Comment by pepyakin** [comment]
Hello!

Yup, you are right, `wabt` doesn't have anything like `ast2wasm` and I agree that the generation of wat is a bit clunky.

I think you might find [`binaryen-rs`](https://github.com/pepyakin/binaryen-rs) useful. [Here](https://github.com/pepyakin/emchipten) is a toy project can help you to get started.

**Comment by justinmchase** [comment]
Ah, looks perfect! I'll check it out! I was looking at the emscripten crate and it wasn't working and seemed like not quite what I was looking for, this looks much better.

**Comment by pepyakin** [comment]
I must say though when I was working with it it wasn't a super pleasant experience. One reason is that binaryen's C API doesn't map nicely on Rust (for instance, it uses `abort` to signal an error condition).

**Comment by justinmchase** [comment]
Yeah, thus is C though. I guess I just won't make any errors!

**Comment by justinmchase** [comment]
Unless you're proposing a from scratch rust implementation? I will be happy to go down the binaryen rabbit hole for a while but it doesn't seem like a far fetched rust project.

**Comment by pepyakin** [comment]
I am not sure if I am proposing anything, but if I were to write a ast→wasm compiler I'd also consider just using raw wasm bytecode (e.g. [parity-wasm](https://github.com/paritytech/parity-wasm)). I am not sure if binaryen-rs would provide big improvements in usability (although it may provide some optimizations. However, you don't have to build in binaryen IR right away, but just can run optimization passes later on on raw wasm),

**Comment by justinmchase** [comment]
Oh yeah, I didn't realize that `parity-wasm` had wasm builder in it, I assumed it was purely for reading and interpreting bytecode... But I suppose it makes sense to have a builder there as its my understanding this is related to ethereum and I supposed you'd need a way to build wasm from one of the various languages. Does binaryen even have any features that this doesn't?

So I'm looking at this sample here:
https://github.com/paritytech/parity-wasm/blob/master/examples/build.rs#L36

And I see its building a body, which appears to be empty. From poking around it looks like it could be:

```.body().with_instructions( /*instructions here */).build()```

Does that sound about right? Is there an example in there of building instructions also? Because yeah that is pretty much exactly what I was looking for. I want to be able to take my AST and then just translate it right into wasm instructions but skip wat.

**Comment by pepyakin** [comment]
> Oh yeah, I didn't realize that parity-wasm had wasm builder in it, I assumed it was purely for reading and interpreting bytecode... But I suppose it makes sense to have a builder there as its my understanding this is related to ethereum and I supposed you'd need a way to build wasm from one of the various languages.

FWIW, parity-wasm doesn't have anything to do with ethereum or any other specific domain. It is a general purpose library which is used in a few projects. For now, I haven't heard about any uses of the builder API.

> Does binaryen even have any features that this doesn't?

Actually, binaryen is a bit different since it works with Binaryen IR which is quite similar but not the same as wasm and because it has quite a lot of optimizations. OTOH parity-wasm is just a serialization/deserialization library. Interpretation has been moved to [wasmi](https://github.com/paritytech/wasmi).

> And I see its building a body, which appears to be empty. From poking around it looks like it could be:

Yep, you can use `with_instructions` like that. I am not sure if there are good examples, but from the back of my mind old tests from pwasm-utils (e.g. [this](https://github.com/paritytech/wasm-utils/blob/a150df8703873b0df8c4779db17230eb0dfd853d/src/gas.rs#L594)) should do.

**Comment by justinmchase** [comment]
Perfect thanks, I think that is exactly what I was looking for! I managed to make a working prototype of what I was trying to do, which is to dynamically generate wasm and then invoke it. I made a little ticket in that project with an example because examples of `with_instructions` were hard to find. That `wasm-utils` function would have done it for sure.

https://github.com/paritytech/parity-wasm/issues/276
